File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxHeight(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return 0
        return 1 + max(self.maxHeight(root.right), self.maxHeight(root.left))

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        leftH = self.maxHeight(root.left)
        rightH = self.maxHeight(root.right)
        if abs(rightH - leftH) > 1:
            return False
        else:
            leftB = self.isBalanced(root.left)
            rightB = self.isBalanced(root.right)
            if leftB == False or rightB == False:
                return False
        return True
            

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* mn = p;
        TreeNode* mx = q;
        
        if (p->val > q->val){
            mn = q;
            mx = p;
        }
        if (root->val >= mn->val && root->val <= mx->val){
            return root;
        }
        else if (root->val > mn->val && root->val > mx->val){
            return lowestCommonAncestor(root->left, mn, mx);
        }
        else 
        {
            return lowestCommonAncestor(root->right, mn, mx);
        }
    }
};
